16_permutations/solution/permutations-solution.js

```diff
@@ -0,0 +1,19 @@
+const permutations = function (original, currentPermutations) {
+  if (original.length === 1) return original;
+
+  const perms = currentPermutations
+    ? currentPermutations
+    : original.map((el) => [el]);
```

Comment:
> ```suggestion
> const permutations = function (original, currentPermutations = original.map((num) => [num])) {
>   if (original.length === 0) return [];
>   if (original.length === 1) return [original];
> ```
> 1. Based on the intent of lines 4-6, instead of the ternary, it'd be clearer to write `const perms = currentPermutations ?? original.map((num) => [num]);`. `||` can be used as well, but the intent of that line really is just if `currentPermutations` doesn't exist, therefore `??` is more semantic. *But*, since that's the only intent of that line, you should just make it a default parameter.
> 1. `el` renamed to `num` to match the instructions. If you wish for the instructions to not be limited to numbers, then `el` can be kept and the instructions amended.
> 1. For an input of `[1]`, it wouldn't make sense to return `[1]` but rather, `[[1]]`. The function returns an array of permutations. The only permutation of `[1]` is `[1]`, so an array containing that would be `[[1]]`, as opposed to `[1]`, which implies the permuted array isn't an array but a number. This is consistent with the other examples, where an array of permuted arrays is returned. The relevant test will need to be updated accordingly.
> 1. If you're going to do a shortcut for 1-length arrays, then you should probably also include one for empty arrays. The above suggestion treats empty arrays as having no permutations, thus returning only an empty array. An alternative, would be to treat an empty array as having exactly one permutation (one way to arrange emptiness). IF that's the case, the two ifs can be combined to just `if (original.length < 2) return [original]`, so that an input of an empty array returns `[[]]`. However you decide to treat empty array inputs, you should make it clear in the instructions how they should be treated, and include a test for that.

1. I like your idea in your proposed code with the default parameters, however I feel like this could be a good case to show the ?? operator, which the learners are unlikely to use and this seems like a good use case of that.

1. I think its best to keep it to numbers not to overcomplicate this exercise, so renaming el to num is a good call

1. Ohh yeah, cant believe I missed that. Thanks!

1. `0!` evaluates to `1`, so a set with no elements will always have a single way to arrange it.

Reply:
Regarding point 1, we really don't need a new variable to maybe be assigned the parameter value. `??` would only be used in that scenario. The equivalent without a new variable would be
```js
partialPermutations ??= original.map((num) => [num]);
```

I'm more in favour of the default param, as I think that's more semantic for this specific scenario (value for parameter if it's initialised `undefined`). I would never personally set a default param via `??=`, I'd rather use the tool meant for that purpose. So personally, I think if we wanted to demonstrate the use of `??=`, it wouldn't be here.

Reply:
> Regarding point 1, we really don't need a new variable to maybe be assigned the parameter value. `??` would only be used in that scenario. The equivalent without a new variable would be
> ```js
> partialPermutations ??= original.map((num) => [num]);
> ```
> 
> I'm more in favour of the default param, as I think that's more semantic for this specific scenario (value for parameter if it's initialised `undefined`). I would never personally set a default param via `??=`, I'd rather use the tool meant for that purpose. So personally, I think if we wanted to demonstrate the use of `??=`, it wouldn't be here.
> 
> Also, `??` is introduced in Foundations somewhere in a javascript.info article, IIRC. I'm sure people will come across it at one point. It's a somewhat recent operator anyway, so it's not like it's replacing anything insanely ridiculous.
> 
> Regarding point 4, nice justification. In that case, we can return `[original]` when the length is `< 2`, then include a test for empty array input, and specify that expectation in the isntructions.

Also, what do you think about if we used `Set` instead of `Array` for this one? I feel like doing that makes *sense* in regards to it, since we often talk about different ways we can arrange a *set* of, consecutive integers in this example.

This would have an added benefit of introducing learners to sets, and this seems to me at least a good use case of that. Often times when we are discussing the permutations of something, we are talking about a set. A deck of cards never has duplicates for example.

Sets are cool and to me feel like a topic small enough to be introduced in an exercise's `README`

Also this would naturally exclude the option with duplicates if we decide to go this route.

Reply:
I'll push back on using Sets instead of arrays.
Sets have their own interface to deal with and while they're no complicated by any means, learning what a Set is and how to use one isn't the point of this exercise, which is instead to recursively generate permutations of something. That should be the sole focus of the challenge.

Reply:
Okay, sure. I understand where you are coming from and sets can be certaintly unnecessary overhead. In that case, I'm going to change this exercise so that instead of taking an array of consecutive integers, it will now take in an array of random integers, which means that there will be tests for duplication. I like your idea (in another comment, but feels like its better to respond here) and I think that certainly makes the challenge more interesting

Also, we could even expand this challenge to allow for all data types like objects, NaN, strings etc. That shouldn't be too complicated as an element that's an array would not have to have the exercise find *that* sub-array's permutations

What do you think?

Reply:
Had a think about this and in all honesty, I think there's no need to deal with different types and duplicates. I think the complexity with just numbers (the current solution already works fine for random non-duplicated numbers so no changes are required) is just right. The main challenge will be how to generate a new permutation (or with your approach, build partial permutations num by num), and I think that should be the primary focus of the task.

Reply:
Alright so I refactored the code and added an additional test for an empty array.